Add custom contact handling

micahhainline opened the pull request:
Added support for issue #100 

As part of WWT Asynchrony Labs integration of CareKit with video chat and CallKit, we needed to have more flexibility in how the Connect page handled contact items.

The new mechanism is flexible at a number of levels, while still providing ease of construction for the OCKContact. While the changes were not entirely backward compatible, the original initializer was maintained (though deprecated) to avoid creating breakages for most users.

Our original hope was to follow the original direction of the OCKContact and expand the usage of API from the Contacts framework (such as CNPhoneNumber) but this turned out not to be practical. The mechanisms that the Contacts application would use to interact with CallKit-enabled applications (launching from bundle ids, querying Intents) are not available to us, as they are private. This left us needing to use custom URL schemes or custom handling of the button presses, both of which we have made into easy options within these changes. Making a clean break from the Contacts framework seems like the best approach going forward.

Icons were added for a new video option, which will default to using Facetime. The default label for this was added as "video" and we made a best-effort attempt to localize this into the supported languages, but it would be reasonable to check our work there.

umerkhan-apple commented on CareKit/Connect/OCKConnectDetailViewController.m:
```diff
@@ -32,7 +33,7 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 #import "OCKDefines_Private.h"
-
+#import "OCKHelpers.h"
 
```
Add extra new line.

umerkhan-apple commented on CareKit/Connect/OCKContact.h:
```diff
@@ -28,10 +29,8 @@
-#import <CareKit/CareKit.h>
 #import <ContactsUI/ContactsUI.h>
-
+#import <CareKit/CareKit.h>
```
?

umerkhan-apple commented on CareKit/Connect/OCKContact.m:
```diff
@@ -32,7 +33,6 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 #import "OCKContact.h"
 #import "OCKHelpers.h"
 
```
Add extra new line.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Add extra new line.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+
```
Add extra new line.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+	OCKContactInfoTypeVideo
+};
+
+@interface OCKContactInfo : NSObject <NSSecureCoding, NSCopying>
```
OCK_CLASS_AVAILABLE

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+	OCKContactInfoTypeMessage,
+	OCKContactInfoTypeEmail,
+	OCKContactInfoTypeVideo
+};
```
Provide documentation for enum. Here's an example:

/**
 An enumeration of the types of contacts available.
 */
OCK_ENUM_AVAILABLE
typedef NS_ENUM(NSInteger, OCKContactType) {
    /**
     A care team contact such as a physician or nurse.
     */
    OCKContactTypeCareTeam = 0,
    
    /**
     A personal contact such as a friend or family member.
     */
    OCKContactTypePersonal
};

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+@property (nonatomic, readonly) NSString *displayString;
+@property (nonatomic, readonly) NSURL *actionURL;
+@property (nonatomic, readonly) NSString *label;
+@property (nonatomic, readonly) UIImage *icon;
```
Provide documentation for all properties of a public class.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+ @param actionURL The action to take to initiate the connection. For example "sms:1-314-555-1234", "tel:1-314-555-1234", "facetime://user@example.com"
+ 
+ @return An instance of the contact info.
+ */
```
Adjust documentation formatting to match CareKit documentation.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+ 
+ @return An instance of the contact info.
+ */
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL;
```
Change `- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL` to `- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(nullable NSURL *)actionURL`.

For consistency.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+ 
+ @return An instance of the contact info.
+ */
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL label:(NSString *)label;
```
Change - (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL to - (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(nullable NSURL *)actionURL.

For consistency.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+ 
+ @return An instance of the contact info.
+ */
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL label:(NSString *)label icon:(UIImage * _Nullable)icon;
```
Refactor nullability.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
++ (OCKContactInfo *)emailContactInfo:(NSString *)emailAddress;
++ (OCKContactInfo *)facetimeVideoContactInfo:(NSString *)emailAddressOrRawPhoneNumber displayString:(NSString * _Nullable)displayString;
++ (OCKContactInfo *)facetimeAudioContactInfo:(NSString *)emailAddressOrRawPhoneNumber displayString:(NSString * _Nullable)displayString;
+
```
Provide documentation for all these class methods. Also refactor nullability as stated above.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.h:
```diff
@@ -0,0 +1,97 @@
+
+@end
+
+NS_ASSUME_NONNULL_END
```
Add extra new line at EOF.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.m:
```diff
@@ -0,0 +1,163 @@
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Add extra new line.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.m:
```diff
@@ -0,0 +1,163 @@
+#import "OCKContactInfo.h"
+#import "OCKHelpers.h"
+#import "OCKDefines_Private.h"
+
```
Add extra new line.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.m:
```diff
@@ -0,0 +1,163 @@
+	return nil;
+}
+
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL {
```
Remove nullability specifiers from implementation files.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.m:
```diff
@@ -0,0 +1,163 @@
+	return [self initWithType:type displayString:displayString actionURL:actionURL label:defaultLabel];
+}
+
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL label:(NSString *)label {
```
Remove nullability specifiers from implementation files.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.m:
```diff
@@ -0,0 +1,163 @@
+	return [self initWithType:type displayString:displayString actionURL:actionURL label:label icon:defaultIcon];
+}
+
+- (instancetype)initWithType:(OCKContactInfoType)type displayString:(NSString *)displayString actionURL:(NSURL * _Nullable)actionURL label:(NSString *)label icon:(UIImage *)icon {
```
Remove nullability specifiers from implementation files.

umerkhan-apple commented on CareKit/Connect/OCKContactInfo.m:
```diff
@@ -0,0 +1,163 @@
+	NSURL *actionURL = [[NSURL alloc] initWithString:[@"facetime-audio://" stringByAppendingString:emailAddressOrRawPhoneNumber]];
+	return [[OCKContactInfo alloc] initWithType:OCKContactInfoTypePhone displayString:displayString ?: emailAddressOrRawPhoneNumber actionURL:actionURL];
+}
+
```
Add extra new line.

umerkhan-apple commented on CareKitTests/CareKitContactTests.m:
```diff
@@ -0,0 +1,109 @@
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Add extra new line.

umerkhan-apple commented on CareKitTests/CareKitContactTests.m:
```diff
@@ -0,0 +1,109 @@
+#import <CareKit/CareKit.h>
+#import <CareKit/CareKit_private.h>
+#import "OCKDefines_Private.h"
+
```
Add extra new line.

umerkhan-apple commented:
@micahhainline Thanks for the PR. This looks great! I've completed a first-pass and provided some feedback.

Will do some device testing and provide further feedback.

micahhainline commented:
@umerkhan-apple Thanks for the quick feedback! Added a new commit correcting all formatting and documentation issues that you pointed out.

umerkhan-apple commented:
@micahhainline There are some merge conflicts due to a recent PR that was merged. https://github.com/carekit-apple/CareKit/pull/78

It added 3D Touch functionality to the Connect module. Could you resolve the conflicts?
Thanks! 👍

micahhainline commented:
@umerkhan-apple We resolved the merge conflicts for 3D Touch. We tested on device to ensure 3D Touch options worked as before.

umerkhan-apple commented:
@micahhainline This looks great! Just 2 last things.

Could you update the sample app and the test app to only use the new initializer? There are several warnings otherwise. 

And could you also update the sample app to include a video contact info item for one of the contacts?

Thanks! With those 2 changes in, this PR will be ready to be merged.

micahhainline commented:
@umerkhan-apple We updated the applications to use the new init method. While doing so we realized some pain points of the long convenience function names, and adjusted them accordingly.
We noticed that some places in the test application are using old Swift syntax, mostly be fully qualifying enums when they don't have to and calling MyClass.init() instead of MyClass(). As we were moving to the new init we fixed those in the classes we touched.

umerkhan-apple commented:
@micahhainline Looks great! Thanks for the quick turnaround! r=me!

micahhainline commented:
@umerkhan-apple Thanks for all the support! Hopefully we'll be able to contribute more to this project in the future.

smravca-apple commented:
Hi @micahhainline! Could you please provide me with a contact email so that I can send you a contribution agreement to sign? Thank you! :)
